dtm0/recovery.c

```diff
@@ -1665,8 +1686,41 @@ static void dtm0_restore(struct m0_fom        *fom,
 		      struct m0_fid       initiator;
 		      struct m0_be_op     reply_op;
 		      bool                next;
```

Comment:
Why is it needed now, why `last_dtx_met` is not enough?

Reply:
I see.

Reply:
Can you add a short comments to these fields, please?

Reply:
does this work?

```diff
diff --git a/dtm0/recovery.c b/dtm0/recovery.c
index 020091a2..b2ad5e2d 100644
--- a/dtm0/recovery.c
+++ b/dtm0/recovery.c
@@ -1729,12 +1729,23 @@ static void dtm0_restore(struct m0_fom        *fom,
        M0_SET0(&F(reply_op));
        m0_be_op_init(&F(reply_op));

+       /*
+        * last_dtx_met and next seem to have very close meaning, but they are
+        * not the same.  last_dtx_met is "we have reached the last transaction
+        * to be analyzed".  next is "we need at least one more iteration" (and
+        * !next is "no more iterations; send EOL message").  The iteration
+        * starts with last_dtx_met=false and next=true.  At the end,
+        * last_dtx_met=true and next=false.  But just before the end, last
+        * entry in the log may or may not need to be sent -- that is when we
+        * need both flags to define what to do.
+        */
        do {
                M0_SET0(&record);
                if (F(last_dtx_met)) {
                        /*
                         * Last iteration reached the RECOVERING mark in the
-                        * log, don't need to send REDOs past this mark.
+                        * log, don't need to send REDOs past this mark.  But we
+                        * still need to send EOL flag (EOL message).
                         */
                        F(next) = false;
                } else {
```

Reply:
Yes, this is good. But still, I'd add a short comment in the structure itself, like these:

```
   ... next; /* redo to be sent present? */
   ...
   ... last_dtx_met; /* stop recovering after this */
```